File: 0338-Counting Bits/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> dp(num+1, 0);
        for(int idx = 1; idx <= num; ++idx) {
            if(idx%2==0) dp[idx] = dp[idx/2];
            else dp[idx] = dp[idx/2]+1;
        } 
        return dp;
    }
};

// Runtime: 4 ms, faster than 98.50% of C++ online submissions for Counting Bits.
// Memory Usage: 8.2 MB, less than 43.61% of C++ online submissions for Counting Bits.

File: 0338-Counting Bits/golang/solution.go
package golang

func countBits(num int) []int {
	dp := make([]int, num+1)
	dp[0] = 0
	for idx := 1; idx <= num; idx++ {
		if idx%2 == 0 {
			dp[idx] = dp[idx/2]
		} else {
			dp[idx] = dp[idx/2] + 1
		}
	}
	return dp
}

// Runtime: 4 ms, faster than 95.54% of Go online submissions for Counting Bits.
// Memory Usage: 4.5 MB, less than 38.22% of Go online submissions for Counting Bits.
